### ISSUE TITLE
[Question] REST API Node - Refused Connection

### ISSUE THREAD

**Issue by callmereno** [thingsboard/thingsboard#7831]
**Component**
REST API Call Node (GET-Request)

 * Rule Engine

**Description**
Hey guys,

i run Thingsboard in a docker container and im trying to perform a REST API GET request, but to the conncetion (to the localhost) gets denied (see 1st picture). Do i maybe have to edit some settings in the Docker Container?

![grafik](https://user-images.githubusercontent.com/114148912/209174695-6a99b6f7-2175-4052-a0dc-e718375d1c11.png)

Here you can see the settings for the node:

![grafik](https://user-images.githubusercontent.com/114148912/209177738-adbae078-0f68-4032-a4bb-1b69dd0b14b4.png)

I looked this issue up and found this proposed solution that didn't work: Set simple client HTTP factory (https://github.com/thingsboard/thingsboard/issues/3434)

Another proposed solution was to use an another URL the locahost, by hosting a free cloud provider. How am I supposed to do that? (https://stackoverflow.com/questions/48702893/thingsboard-i-o-error-on-post-request)

I also tried the same REST API Call via the TB-Demoserver, but i told me that the authentifcation failed (401). What could i do to make it work there?

Any ideas to solve this issue would be much appreciated.

Greetings
Reno

**Environment**

 * OS:  Windows 10
 * ThingsBoard: v.3.4.1 (Docker)
 * Browser: Firefox 108.0.1 (64-Bit)

**Comment by kastayou** [comment]
API authentication is not basic authentication but a JWT token authentication.
https://thingsboard.io/docs/reference/rest-api/

You need to make a first api call with your username and password to get the JWT token and then use this token to make the get timeseries keys request.

**Comment by callmereno** [comment]
How do make thta first API call? Do I insert another API node in the Rule Engine?
I have copied the JWT Token from Settings > Security. Can i do something with that?

**Comment by callmereno** [comment]
I tried it with this Curl:


curl -X POST --header 'Accept:application/json' -d '{\"username\":"tenant@thingsboard.org\",\"password\":\"tenant\"}' http://localhost:8080/api/auth/login

But CMD tells me that no URL is specified.

@kastayou:

**Comment by devaskim** [comment]
You could use [this rule chain](https://github.com/devaskim/awesome-thingsboard/tree/main/chain/tb_api_token) in order to login to Thingsboard and get JWT token for further REST requests to TB API

**Comment by callmereno** [comment]
Hey @devaskim,

this looks very promicing. The problem is that the 'login' rest api node doesn't seem to use login credentails it gets from the 'login' script node:

![grafik](https://user-images.githubusercontent.com/114148912/209329024-dcc93096-c815-4671-931d-89a0ed1349c2.png)

I added this rule chain after the message type switch in the root rule chain and infuse the telemetry data. This shouldnt be the issue right? 




I tried posting the credentials and getting the JWT tokens via swagger and it worked fine as you can see:

![grafik](https://user-images.githubusercontent.com/114148912/209325559-116cfbb5-5bcc-4b7f-be1f-5cf34809814c.png)

That gave me the idea to add the the 1st JWT token from the image above as header in my GET API Call: `X-Authorization: Bearer ...`:

![grafik](https://user-images.githubusercontent.com/114148912/209330165-9ed94946-6ab6-4390-abc0-b26ceec723ea.png)

But i get the similar error: 

![grafik](https://user-images.githubusercontent.com/114148912/209330442-335a0829-cb6f-420e-8f4c-4aa3e73100cd.png)

Thanks for your help so far!

**Comment by devaskim** [comment]
> This shouldnt be the issue right?

Right.

`Connection refused` is network issue not TB. Unfortunately, I have no idea why this happens with REST rule node, but not with Swagger API

**Comment by callmereno** [comment]
> `Connection refused` is network issue not TB

Yeah that's something i assumed aswell. I will try to replecate the Rule Chains on the demoserver and see what happens.

**Comment by callmereno** [comment]
@kastayou Do you think i could have something to do with me running TB in a docker container?

**Comment by devaskim** [comment]
Well, in my Windows 10 case I did nothing special. Just download and run

**Comment by callmereno** [comment]
Can you maybe show me what happens in your login api node?

**Comment by devaskim** [comment]
My bad, login node works on my pc with external IP not localhost. Sorry.

**Comment by callmereno** [comment]
hey @devaskim 

Is there a possbilty to setup an external IP für the docker instance?

**Comment by devaskim** [comment]
Well, I am deploying TB in docker on VPS (i.e. with external IP) and I didn't make any specific configuration

**Comment by callmereno** [comment]
Okay i see. It will consider this for the future :)

I tried the rest api call with demothingsboard and it seems to work, since i get an other error now:

org.springframework.web.client.HttpServerErrorException$InternalServerError: 500 : "{"status":500,"message":"Failed to convert value of type 'java.lang.String' to required type 'java.lang.Long'; nested exception is java.lang.NumberFormatException: For input string: \"ss_startInt\"","errorCode":2,"timestamp":"2022-12-27T12:59:50.206+00:00"}"

ss_startInt is a value that you can adjust on a dashboard and is stored as string in the metadata. I know that i can reference it with ${ss_startInt} in the url itself. But how do i parse it into the Long-format. I've only found the parseInt()-function in TBEL documentation and it doesn't even work. This is how the script (String to Long) i want to use looks right now:

![grafik](https://user-images.githubusercontent.com/114148912/209671102-220ba43f-840a-4c5a-af82-aa9abaea1c74.png)

As you can see i tried to parse "ss_endInt", but it didnt change the payload. I'm very new in this topic as you might have noticed but any help is very appreiciated. 

Thanks for the quick response btw.

**Comment by devaskim** [comment]
It is JS related issue not TB
`metadata.ss_endTs = parseInt(metadata.ss_endTs, 10;)`
or just
``metadata.ss_endTs = +metadata.ss_endTs;`

**Comment by callmereno** [comment]
I used Long.valueOf(), since parseInt() gave out a value which was diffrent from the actual timestamp value.
Also i put the values into the msg, since metadata can only hold string values. Is that correct

![grafik](https://user-images.githubusercontent.com/114148912/209814879-524cf5c1-4851-4a6d-a5c6-e8529b3c8e88.png)

Thank you, for the help :)

**Comment by devaskim** [comment]
Where to put parsing results depends only on your needs
